### ISSUE TITLE
Custom MediaDescriptionAdapter ContentTitle being Overloaded elsewhere

### ISSUE THREAD

**Issue by AnthonyMarkD** [google/ExoPlayer#9857]
Hello, While working with Exoplayer for media playing I have come across a weird behaviour where the MediaDescriptionAdapter method of getCurrentContentTitle() seems to be overloaded elsewhere in the code.

When creating a notification I create my own Description adapter like so, and I can observe it working as it overloads the Large Icon the description and the content Intent. However, the content title does not get overloaded and instead it uses the media Item's title.

```
private inner class DescriptionAdapter(private val controller: MediaControllerCompat) : PlayerNotificationManager.MediaDescriptionAdapter {

        var currentIconUri: Uri? = null
        var currentBitmap: Bitmap? = null

        override fun getCurrentLargeIcon(
            player: Player,
            callback: BitmapCallback
        ): Bitmap? {
            val iconUri = controller.metadata.description.iconUri
            return if (currentIconUri != iconUri || currentBitmap == null) {

                currentIconUri = iconUri
                scope.launch {
                    currentBitmap = iconUri?.let {
                        resolveUriAsBitmap(it)
                    }
                    currentBitmap?.let { callback.onBitmap(it) }
                }
                null
            } else {
                currentBitmap
            }
        }

        override fun getCurrentContentTitle(player: Player): CharSequence {
            Log.d("Notification2", controller.metadata.description.title.toString() + "asd")
            return controller.metadata.description.title.toString() + "asd"
        }


        override fun createCurrentContentIntent(player: Player): PendingIntent? {
            return controller.sessionActivity
        }

        override fun getCurrentContentText(player: Player): CharSequence? {
            return controller.metadata.description.subtitle.toString() + "Hi"
        }

        private suspend fun resolveUriAsBitmap(uri: Uri): Bitmap? {
            return withContext(Dispatchers.IO) {
                // Block on downloading artwork.
                val request = ImageRequest.Builder(this@MediaPlaybackService)
                    .data(uri)
                    .size(144, 144)
                    .build()
                val drawable = this@MediaPlaybackService.imageLoader.execute(request).drawable
                return@withContext drawable?.toBitmap()
            }
        }
    }
```

When i disabled my Queue navigator, the content is properly overloaded, hence the asd is shown at the end of the notification title, but if I have it enabled, It fails to show it. This only happens to the title as well for some reason.

```
 mediaSessionConnector.setQueueNavigator(object : TimelineQueueNavigator(mediaSession!!) {
            override fun getMediaDescription(
                player: Player,
                windowIndex: Int
            ): MediaDescriptionCompat {
                Log.d("mediaItemAtIndex", windowIndex.toString())
                val mediaItemAtIndex = player.getMediaItemAt(windowIndex)

                return MediaDescriptionCompat.Builder()
                    .setTitle("hello")
                    //.setTitle(mediaItemAtIndex.mediaMetadata.title)
                    .setDescription(mediaItemAtIndex.mediaMetadata.description)
                    .setSubtitle(mediaItemAtIndex.mediaMetadata.artist)
                    .setIconUri(mediaItemAtIndex.mediaMetadata.artworkUri)
                    .build()
            }
        })
```

![Screenshot_20220112-182338_VOM](https://user-images.githubusercontent.com/44737462/149249164-ff8f11f0-515d-41e5-a17e-a8a62601589f.jpg)
![Screenshot_20220112-182405_VOM](https://user-images.githubusercontent.com/44737462/149249170-ac2384a5-2fac-45a4-9a23-710aa131e35d.jpg)

As the screen shots show, there is no asd at the end of the content title, but there is a Hi for the content text, and I am completely confused as to why.

- ExoPlayer version number 2.16.1
 Android 11
 Samsung s20 Plus

**Comment by icbaker** [comment]
Please provide a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) that demonstrates the problem in a way that we can build locally. 

This could be an Android Studio project on GitHub, or zipped up and sent to dev.exoplayer@gmail.com using a subject in the format "Issue #1234", where "#1234" should be replaced with your issue number. Please also update this issue to indicate you’ve done this.

**Comment by AnthonyMarkD** [comment]
Here is a [minimal reproducible sample](https://github.com/AnthonyMarkD/AudioPlayerSampleTest) Which plays an audio file off the internet. As seen, the Description Adapter does not overload the content title and text with strings, instead choosing to use the media title and artist like the default one does, but it does overload the image.

**Comment by icbaker** [comment]
> As seen, the Description Adapter does not overload the content title and text with strings, instead choosing to use the media title and artist like the default one does, but it does overload the image.

Thanks for the project. I ran it, but I'm not sure if I'm seeing the same behaviour as you.

I've attached a screenshot showing the mini-player I see on Android 11 on a OnePlus AC2003. It seems to show all the information that your app hard-codes into `MediaMetadata` in `onStartCommand` (not just the image, like your description suggests).

Could you please provide precise details about exactly what you expect to see with your example project and what you actually see (including a screenshot)?

![android-screenshot-2022-01-24T170929](https://user-images.githubusercontent.com/1279752/150830871-263bc7f9-6367-41e3-b2e6-8d42d87d90a3.png)

**Comment by AnthonyMarkD** [comment]
This might be due to me misunderstanding what each component does in the library, but to my understanding, you can provide a custom MediaDescriptionAdapter to help supplement media content to the PlayerNotificationManager. In my code, I have the description adapter return a hardcoded string "tested" for the content title, and "hello" for the content text. Why does the notification not show these 2 values, and instead shows the hardcoded media property values.

The screenshot is what appears on a Samsung Galaxy Plus s20 , which show to have the same content as yours.

![Screenshot_20220124-104115_AudioPlayerTest](https://user-images.githubusercontent.com/44737462/150835896-aaa0ebef-3bab-4103-9897-abdc7a83bddf.jpg)

**Comment by icbaker** [comment]
Ah I understand - thanks for clarifying.

I think what's happening here is related to the changes to media notification/players in Android 11: https://developer.android.com/guide/topics/media/media-controls

The mini-player in Android 11+ uses metadata directly from the media session, and is therefore unaffected by the 'overrides' in your `DescriptionAdapter`. I tried your app on an Android 10 device (original Pixel) and it shows the 'overridden' "tested" and "hello" strings because it's showing a 'notification' rather than the mini-player.

I'm going to pass this issue over to @marcbaechinger as he's more familiar with the details of notifications and media sessions - and can hopefully add more detail or correct me if I've got anything wrong!

![android-screenshot-2022-01-24T182742](https://user-images.githubusercontent.com/1279752/150842276-bd38304a-6f7f-4ffe-af95-9c465c1de26c.png)

**Comment by marcbaechinger** [comment]
That's right, you need to make sure that you return the same values with `PlayerNotificationManager.MediaDescriptionAdapter` and what is passed to `MediaSession.setMetadata(MediaMetadataCompat)`. That's the only way to get a consistent picture on all API levels for notifications as well as mini player (also for lock screen notifications and clients like Android Auto/Wear OS to be consistent with what you show in the notification).

The latter is dependent on whether you are using `TimelineQueueNavigator` and/or you are using a custom `MediaMetadataProvider`:

If you are using the most common approach that is `TimelineQueueNavigator` with `DefaultMediaMetadataProvider` you need to make sure to return the proper values with what you return from the abstract method `TimelineQueueNaviogator.getMediaDescription(Player player, int windowIndex)` that you have implemented. If you are using a custom `MediaMetadataProvider` that you have set with `MediaSessionConnector.setMediaMetadataProvider(MediaMetadataProvider)` then you need to make sure to return the proper values when the provider is called.

**Comment by AnthonyMarkD** [comment]
Ahh that makes sense, thank you very much for the clarification!
